ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Node/DiffusionPipelining.hs

```diff
@@ -1,16 +1,117 @@
-{-# LANGUAGE DerivingVia          #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE DeriveAnyClass       #-}
+{-# LANGUAGE DeriveGeneric        #-}
+{-# LANGUAGE DerivingStrategies   #-}
 {-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE LambdaCase           #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
 {-# LANGUAGE StandaloneDeriving   #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeFamilies         #-}
 {-# LANGUAGE UndecidableInstances #-}
 
 {-# OPTIONS_GHC -Wno-orphans #-}
 
-module Ouroboros.Consensus.Shelley.Node.DiffusionPipelining () where
+module Ouroboros.Consensus.Shelley.Node.DiffusionPipelining (
+    HotIdentity (..)
+  , ShelleyTentativeHeaderState (..)
+  ) where
 
+import qualified Cardano.Ledger.Api.Era as L
+import qualified Cardano.Ledger.Shelley.API as SL
+import           Control.Monad (guard)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Word
+import           GHC.Generics (Generic)
+import           NoThunks.Class
 import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Shelley.Eras (StandardConway)
 import           Ouroboros.Consensus.Shelley.Ledger.Block
 import           Ouroboros.Consensus.Shelley.Ledger.Protocol ()
+import           Ouroboros.Consensus.Shelley.Protocol.Abstract
 
-deriving via SelectViewDiffusionPipelining (ShelleyBlock proto era)
-  instance ShelleyCompatible proto era
-  => BlockSupportsDiffusionPipelining (ShelleyBlock proto era)
+-- | Hot block issuer identity for the purpose of Shelley block diffusion
+-- pipelining.
+data HotIdentity c = HotIdentity {
+    -- | Hash of the cold key.
+    hiIssuer  :: !(SL.KeyHash SL.BlockIssuer c)
+  , -- | The issue number/opcert counter. Even if the opcert was compromised and
+    -- hence an attacker forges blocks with a specific cold identity, the owner
+    -- of the cold key can issue a new opcert with an incremented counter, and
+    -- their minted blocks will be pipelined.
+    hiIssueNo :: !Word64
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving anyclass (NoThunks)
+
+data ShelleyTentativeHeaderState proto =
+    -- | Legacy state, can be removed once mainnet is in Conway.
+    LegacyShelleyTentativeHeaderState !(SelectViewTentativeState proto)
+  | ShelleyTentativeHeaderState
+      -- | The block number of the last trap tentative header.
+      !(WithOrigin BlockNo)
+      -- | The set of all hot identies of those who issued trap tentative
```

Comment:
I suggest clarifying that the scope of "all ... who ..." is dependent on the context. In ChainSel, it's all the blocks this node has discovered are invalid. In BlockFetch clients, it's what the upstream peer has sent. Right?

Reply:
Expanded on this :+1: